### DISCUSSION TITLE
Updating state in par_for_each

### DISCUSSION THREAD

**Discussion by rks-malvern** [bevyengine/bevy#6979] [Q&A] [answered]
Howdy,

Rust & Bevy newbie here. I want to play around with an architecture that does a lot of compute-heavy work (AI/strategy, pathfinding) over an entirely read-only set of data, so that the system can lookup data from other entities while retaining full parallelization.

Because I want to access other entities, I can't update anything in the closure. Therefore, I was going to use `Commands` to issues updates. But `mut Commands` aren't `Clone`, so I can't pass them to the closure.

```
fn compute_heavy_system(q: Query<Entity, &Comp1>, mut cmds: Commands) {
    q.par_for_each(BATCH, |ent, comp1| {
        if let Ok(other_ent, other_comp) = q.get(comp1.target_ent) {
            // Do computations
            // comp1.strategy = XXX;    // ERR: must be immutable for to allow q.get()
            // cmds.get_entity(ent);    // ERR: cmds isn't Clone so can't be sent to thread
        }
    }
}
```
It strikes me that there is no fundamental reason that Commands can't be issued in the workers; either
by making Commands synchronized or by having thread-local Commands.

I thought the tradeoffs of this architecture could be interesting:
  + PRO: access to wide variety of data allows complex decision making
  + PRO: flexibility in data access eliminates data restructuring when new logic is added
  + PRO: full parallelization possible
  + CON: depending on how much non-row data is needed, worsens data locality in system
  + CON: no ordering guarantees on updates in a frame
  + CON: updates are serialized and async (slower, sometimes surprising)

Am I missing something? Is my architecture obviously bad? If so, why, and what are alternatives?

**Comment by soqb** [comment, marked as answer]
there is https://docs.rs/bevy/latest/bevy/prelude/struct.ParallelCommands.html

**Reply by rks-malvern** [reply]
Thanks! How did I miss that?

**Reply by james7132** [reply]
Note: there's also `Query::get_unchecked_mut`, but it's unsafe and you need to prove it never aliases. If you can't, it's best to use `ParallelCommands` instead.
